### DISCUSSION TITLE
Getting rate limited on GitHub but don't know why

### DISCUSSION THREAD

**Discussion by mrsimo** [renovatebot/renovate#19647] [Request Help]
Hi!

We run renovate ourselves, passing a GitHub access token via `RENOVATE_TOKEN` env var. We run it a few times an hour against multiple repos. Renovate generally works, but we see a problem: we have an internal package that every time it gets released Renovate should create 10+ PR's in different repositories, but it sometimes takes multiple runs of Renovate for it to open the PR's in all the repos

We've seen a few hundred errors a week saying `Rate limit exceeded - aborting` for specific repos. I'm not mentioning exact numbers because it varies quite a bit.

We setup monitoring on the access token's rate limits and they never really go anywhere near 0 remaining:

<img width="1193" alt="image" src="https://user-images.githubusercontent.com/19203/210346326-d525a7fa-1557-44b5-9ca9-4bba19cdd14e.png">

In the same period of time we had 810 `Rate limit exceeded - aborting` errors.

Looking into more detail I see:

<img width="893" alt="image" src="https://user-images.githubusercontent.com/19203/210347365-8b32dbcb-fd5c-451e-9671-94f30e89f24a.png">

Or:

<img width="893" alt="image" src="https://user-images.githubusercontent.com/19203/210347477-2980f464-50bd-4c3d-8e1b-e812e8707379.png">

Are we missing something here? Any advice would be welcome.

**Comment by rarkins** [comment]
Google up "github secondary rate limit" for details. It's not the same as the primary rate limit which you're looking at. If you are an enterprise customer you could try raising a support request asking if there's anything they can do to raise your limits

**Reply by mrsimo** [reply]
Ah, that was obvious from the error, wasn't it? I thought I'd gone through all the github docs on rate limits, but clearly missed that. I noticed more information in one of the logs, but they don't say anything useful other than they're secondary rate limits.

We're going to try limiting concurrency or requests per second, see if it helps. Thanks a lot for the response.

**Reply by HonkingGoose** [reply]
> I noticed more information in one of the logs, but they don't say anything useful other than they're secondary rate limits.

Can we edit our debug message so it explains where to find the official GitHub Docs for secondary rate limits?

**Reply by mrsimo** [reply]
In case it's useful for someone else, the rate limit errors seem to have stopped with these settings:

```json
"hostRules": [
  {
    "matchHost": "api.github.com",
    "concurrentRequestLimit": 1,
    "maxRequestsPerSecond": 8
  }
]
```

Run times have doubled from ~5m to ~10m for 28 repos, which is still quite acceptable, and we can keep running more than once an hour.

**Reply by mrsimo** [reply]
> Can we edit our debug message so it explains where to find the official GitHub Docs for secondary rate limits?

From all these log lines:

<img width="849" alt="image" src="https://user-images.githubusercontent.com/19203/211279109-1c782b1a-16fe-4e7a-b869-a825003cb089.png">

The one that says "GitHub failure: secondary rate limit" has all the response information from the request, which does include:

<img width="1211" alt="image" src="https://user-images.githubusercontent.com/19203/211279365-a9644777-ccb8-4eb9-abc8-f88e00b14e39.png">

Which does point to [the docs](https://docs.github.com/en/free-pro-team@latest/rest/overview/resources-in-the-rest-api#secondary-rate-limits), and that one links to the best practices one, but that log line is a `debug` level one. In fact, the only non-debug log is a `warn` that says `Rate limit exceeded - aborting`.

The logs seem to imply a secondary rate limit is detected (so Renovate knows what it is), pointing to those logs might help someone in the future, so 👍🏻

**Reply by rarkins** [reply]
Please add a feature request issue for Renovate to log a warning about GitHub secondary rate limits

**Reply by HonkingGoose** [reply]
I created an issue to track this idea:

- https://github.com/renovatebot/renovate/issues/19746
